modules/packages/OrderedSet.chpl

```diff
@@ -0,0 +1,484 @@
+
+    :arg a: An orderedSet to take the union of.
+    :arg b: An orderedSet to take the union of.
+
+    :return: A new orderedSet containing the union between `a` and `b`.
+    :rtype: `orderedSet(?t)`
+  */
+  proc |(const ref a: orderedSet(?t), const ref b: orderedSet(t)): orderedSet(t) {
+    var result: orderedSet(t, (a.parSafe || b.parSafe));
+
+    result = a;
+    result |= b;
+
+    return result;
+  }
+
+  /*
+    Add to the orderedSet `lhs` all the elements of `rhs`.
+
+    :arg lhs: An orderedSet to take the union of and then assign to.
+    :arg rhs: An orderedSet to take the union of.
+  */
+  proc |=(ref lhs: orderedSet(?t), const ref rhs: orderedSet(t)) {
+    for x in rhs do
+      lhs.add(x);
+  }
+
+  /*
+    Return a new orderedSet that contains the union of two sets. Alias for the `|`
+    operator.
+
+    :arg a: An orderedSet to take the union of.
+    :arg b: An orderedSet to take the union of.
+
+    :return: A new orderedSet containing the union between `a` and `b`.
+    :rtype: `orderedSet(?t)`
+  */
+  proc +(const ref a: orderedSet(?t), const ref b: orderedSet(t)): orderedSet(t) {
+    return a | b;
+  }
+
+  /*
+    Add to the orderedSet `lhs` all the elements of `rhs`.
+
+    :arg lhs: An orderedSet to take the union of and then assign to.
+    :arg rhs: An orderedSet to take the union of.
+  */
+  proc +=(ref lhs: orderedSet(?t), const ref rhs: orderedSet(t)) {
+    lhs |= rhs;
+  }
+
+  /*
+    Return a new orderedSet that contains the difference of two sets.
+
+    :arg a: An orderedSet to take the difference of.
+    :arg b: An orderedSet to take the difference of.
+
+    :return: A new orderedSet containing the difference between `a` and `b`.
+    :rtype: `orderedSet(t)`
+  */
+  proc -(const ref a: orderedSet(?t), const ref b: orderedSet(t)): orderedSet(t) {
+    var result = new orderedSet(t, (a.parSafe || b.parSafe));
+
+    for x in a do
+      if !b.contains(x) then
+        result.add(x);
+
+    return result;
+  }
+
+  /*
+    Remove from the orderedSet `lhs` the elements of `rhs`.
+
+    .. warning::
+
+      This will invalidate any references to elements previously contained in
+      the orderedSet `lhs`.
+
+    :arg lhs: An orderedSet to take the difference of and then assign to.
+    :arg rhs: An orderedSet to take the difference of.
+  */
+  proc -=(ref lhs: orderedSet(?t), const ref rhs: orderedSet(t)) {
+    for x in rhs do
+      lhs.remove(x);
+  }
+
+  /*
+    Return a new orderedSet that contains the intersection of two sets.
+
+    :arg a: An orderedSet to take the intersection of.
+    :arg b: An orderedSet to take the intersection of.
+
+    :return: A new orderedSet containing the intersection of `a` and `b`.
+    :rtype: `orderedSet(t)`
+  */
+  proc &(const ref a: orderedSet(?t), const ref b: orderedSet(t)): orderedSet(t) {
+    var result: orderedSet(t, (a.parSafe || b.parSafe));
+
+    /* Iterate over the smaller orderedSet */
```

Comment:
These various operations can be improved when Chapel has iterators like C++. Or we have to move procedures down to the underlying implementation to optimize. In the second case, the operation can only be applied on OrderedSet with the same implementation because each implementation doesn't know about others. I don't like that so just use the naive method and wait for more powerful iterator support in Chapel.

Reply:
I agree with this in general. But I am not sure what exactly is missing in Chapel's iterators today. Can you elaborate more on that? Maybe with a code example that you wish it worked in Chapel today but it doesn't?

Reply:
Something like this:
```chpl
var a, b, c: orderedSet;
var ait = a.getIter();
var bit = b.getIter();
while ait != nil {
  while bit != nil && ait.value() > bit.value() then bit = bit.next();
  if bit == nil then break;
  else if bit.value() == ait.value() then c.insert(ait.value());
  ait = ait.next();
}
```
There needs a way to have more control over how iterators move forward.

Reply:
Oh, I see what you are saying.